libs/@local/harpc/server/src/serve.rs

```diff
@@ -0,0 +1,46 @@
+use futures::{Stream, StreamExt};
+use harpc_net::session::server::Transaction;
+use harpc_tower::{
+    body::server::request::RequestBody,
+    net::pack::Pack,
+    request::{self, Request},
+    response::BoxedResponse,
+};
+use tokio::pin;
+use tower::{Service, ServiceExt};
+
+// TODO: do we want `BoxedResponse` to be `!`? or is there a better way of doing this? We could have
+// a body that takes any E and converts it to a `!` error (by adding a `TransactionError` and
+// terminating)
+pub async fn serve<M, S>(stream: impl Stream<Item = Transaction> + Send, mut make_service: M)
+where
+    M: Service<(), Response = S, Error = !, Future: Send> + Send,
+    S: Service<Request<RequestBody>, Response = BoxedResponse<!>, Error = !, Future: Send>
+        + Send
+        + 'static,
+{
+    pin!(stream);
+
+    #[expect(
+        clippy::significant_drop_in_scrutinee,
+        reason = "Semaphore permit being dropped is expected, used for control flow in Arc"
+    )]
+    while let Some(transaction) = stream.next().await {
+        let (context, sink, stream) = transaction.into_parts();
+
+        let parts = request::Parts::from_transaction(&context);
+        let request = Request::new(parts, RequestBody::new(stream));
+
+        let Ok(service): Result<S, !> = make_service.call(()).await;
+
+        tokio::spawn(async move {
+            let Ok(response): Result<BoxedResponse<!>, !> = service.oneshot(request).await;
+            let response = response.into_body();
+
+            let pack = Pack::new(response).map(Ok);
+            if let Err(error) = pack.forward(sink).await {
+                tracing::error!(?error, "failed to send response");
+            }
+        });
```

Comment:
What is happening to the handles? Shouldn't we at least return a drop-guard so the threads are properly joined?

Reply:
see: [e876238](https://github.com/hashintel/hash/pull/5320/commits/e876238c04a3905235eb8c066d4115ee31792859)